### PR TITLE
Helm k8s openshift

### PR DESCRIPTION
**Pull request by zubcevic**
This pull request adds an additional deployment option. It discusses how to use the helm chart for openshift and provides sample for enriching the images with a custom certificate authority and other required changes for a strict rootless and random user id k8s environment

### CODE REVIEW

File: openshift/dockerfiles/README.md [hunk 1]
```diff
@@ -0,0 +1,49 @@
+# Docker - Podman rebuild for Enterprises 
+
+For companies that use their own certificate authorities, the Dockerfiles supplied in this directory can be used as an example to add ca certificates to trust stores.
+Also the proxy image is adjusted so it will be able to run as non-root on openshift.
+
+## podman build
+
+You can build the images from the ICTU base images. so either pull does images or first build these images. Then you can build and tag your extensions.
```
Comment (fniessink) [inline openshift/dockerfiles/README.md]:
```suggestion
You can build the images from the ICTU base images. so either pull those images or first build these images. Then you can build and tag your extensions.
```

File: openshift/dockerfiles/README.md [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+	podman build -t containers.local/somenamespace/quality-time_proxy:v4.5.0 --build-arg IMAGE_NAME=registry.access.redhat.com/ubi8/nginx-120 -f Dockerfile.proxy .
+	podman build -t containers.local/somenamespace/quality-time_collector:v4.5.0 --build-arg IMAGE_NAME=docker.io/ictu/quality-time_collector:v4.5.0 -f Dockerfile.collector .
+	# and more builds followed by login and pushes to containers.local repo
+	
+Make sure to adjust the [docker-compose.yml](../../docker/docker-compose.yml) or the helm chart [values.yaml](../helm/values.yaml) with your extension images (such as containers.local/somenamespace/quality-time_collector:v4.5.0).
+	
+## Changes for custom Certificate Authorities
+
+If your organization uses their own certificate authority, you need to add the certificate to the correct trust stores in the images that have connections to the resources that have certificates that were issued by this certificate authority.
```
Comment (fniessink) [inline openshift/dockerfiles/README.md]:
```suggestion
If your organization uses its own certificate authority, you need to add the certificate to the correct trust stores in the images that have connections to the resources that have certificates that were issued by this certificate authority.
```

File: openshift/dockerfiles/README.md [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+    USER root
+    COPY *.crt /usr/local/share/ca-certificates/
+    COPY *.crt /etc/ssl/certs/ 
+    RUN /usr/sbin/update-ca-certificates
+    RUN cd /usr/local/share/ca-certificates/ ; for key in $(ls *.crt) ;do cat $key >> /usr/lib/ssl/cert.pem ; done 
+    ENV REQUESTS_CA_BUNDLE /usr/lib/ssl/cert.pem
+
+[Dockerfile.externalserver](./Dockerfile.externalserver) show an example to add the certificate and enable it at OS level and Python application level.
```
Comment (fniessink) [inline openshift/dockerfiles/README.md]:
```suggestion
[Dockerfile.externalserver](./Dockerfile.externalserver) shows how to add the certificate and enable it at OS level and Python application level.
```

File: openshift/dockerfiles/README.md [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+    RUN chgrp -R 0 /home/collector && chmod -R g=u /home/collector
+	
+[Dockerfile.collector](./Dockerfile.collector) shows the fix for this in the collector image.
+
+
+
+
+	
+
```
Comment (fniessink) [inline openshift/dockerfiles/README.md]:
```suggestion
```

File: openshift/dockerfiles/nginx.conf [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+}
+
+
+
+
+
+
+
+
```
Comment (fniessink) [inline openshift/dockerfiles/nginx.conf]:
```suggestion
```

File: openshift/dockerfiles/nginx.conf [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+        # Load configuration files for the default server block.
+        include /opt/app-root/etc/nginx.default.d/*.conf;
+
+        location /api/v3/nr_measurements {
+			gzip                off;
+			proxy_http_version  1.1;
+			proxy_read_timeout  24h;
+			proxy_set_header    Connection "";
+			proxy_pass          http://externalserver:5001/api/v3/nr_measurements;
```
Comment (fniessink) [inline openshift/dockerfiles/nginx.conf:61]: In the docker compose we use `external_server` as server name. Is there a specific reason why it's `external server` here? If not, I'd prefer to keep the service names the same.

File: openshift/helm/Chart.yaml [hunk 7]
```diff
@@ -0,0 +1,6 @@
+apiVersion: v2
+name: quality-time
```
Comment (fniessink) [inline openshift/helm/Chart.yaml:2]:
```suggestion
name: Quality-time
```

File: openshift/helm/Chart.yaml [hunk 8]
```diff
@@ -0,0 +1,6 @@
...
+description: A Helm Chart for running ICTU quality-time on OpenShift
```
Comment (fniessink) [inline openshift/helm/Chart.yaml:3]:
```suggestion
description: A Helm Chart for running Quality-time on OpenShift
```

File: openshift/helm/README.md [hunk 9]
```diff
@@ -0,0 +1,29 @@
+# Helm chart
+
+A Helm chart allows for a simple deployment on a Kubernetes cluster. The Helm chart in this directory is based on the docker-compose.yml and the standard images and settings from ICTU.
+Except for the Route these are all standard kubernetes elements. The Route defines the main entry point for the application on OpenShift. It is added as the last element in the [templates/www-service.yaml](templates/www-service.yaml)
```
Comment (fniessink) [inline openshift/helm/README.md]:
```suggestion
Except for the Route these are all standard kubernetes elements. The Route defines the main entry point for the application on OpenShift. It is added as the last element in the [templates/www-service.yaml](templates/www-service.yaml).
```

File: openshift/helm/README.md [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+Also in a corporate environment, it is usually required to define credentials for pulling the images. For this a [templates/pull-secret.yaml](templates/pull-secret.yaml) file is provided.
+All templates refer to standard values from [values.yaml](values.yaml) uneless these are overriden using the Helm way of value overrides. (Either by defining seperate values or your own values.yaml file).
```
Comment (fniessink) [inline openshift/helm/README.md]:
```suggestion
All templates refer to standard values from [values.yaml](values.yaml) unless these are overridden using the Helm way of value overrides. (Either by defining seperate values or your own values.yaml file).
```

File: openshift/helm/README.md [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+
+## Pre-requisites
```
Comment (fniessink) [inline openshift/helm/README.md]:
```suggestion
## Prerequisites
```

File: openshift/helm/README.md [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
++ Helm the Helm CLI
+
+Login with the OC CLI and select a project/namespace. Then the helm commands can be used.
+
+## Install
+
+    helm install quality-time .
+    helm install --set pullsecret=mybase64encodeddockerjson quality-time .
+    helm install --values myoverride-values.yaml quality-time . 
+
```
Comment (fniessink) [inline openshift/helm/README.md]:
````suggestion
```console
$ helm install quality-time .
$ helm install --set pullsecret=mybase64encodeddockerjson quality-time .
$ helm install --values myoverride-values.yaml quality-time . 
```
````

File: openshift/helm/README.md [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+
+## Uninstall
+
+    helm uninstall quality-time
```
Comment (fniessink) [inline openshift/helm/README.md]:
````suggestion
```console
$ helm uninstall quality-time
```    
````

File: openshift/helm/templates/pull-secret.yaml [hunk 14]
```diff
@@ -0,0 +1,14 @@
...
+metadata:
+  name: quality-time-pullsecret
+type: kubernetes.io/dockerconfigjson
+---
+apiVersion: v1
+imagePullSecrets:
+- name: quality-time-pullsecret
+kind: ServiceAccount
+metadata:
+  name: quality-time-sa
```
Comment (fniessink) [inline openshift/helm/templates/pull-secret.yaml]:
```suggestion
  name: quality-time-sa
  
```

File: openshift/helm/values.yaml [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+renderer:
+  name: renderer
+  image: docker.io/ictu/quality-time_renderer:v4.5.0
+
+www:
+  name: www
+  image: docker.io/ictu/quality-time_proxy:v4.5.0
+
+
+
```
Comment (fniessink) [inline openshift/helm/values.yaml]:
```suggestion
```

**Review by fniessink — CHANGES_REQUESTED** [review]
Hi René, thanks again for the PR. I've flagged a few small issues. Also one request, could you add a README in the openshift folder briefly explaining the contents of the folder? Or maybe move the openshift/helm/README.md to openshift/?

File: openshift/dockerfiles/nginx.conf (hunk 6 above)
Comment (zubcevic) [reply openshift/dockerfiles/nginx.conf:61]: The underscore is not allowed for service connection in OpenShift. Not sure if it is company specific or a Red Hat standard

File: openshift/helm/Chart.yaml (hunk 7 above)
Comment (zubcevic) [reply openshift/helm/Chart.yaml:2]: The name is based on the naming convention for Helm charts: https://helm.sh/docs/chart_best_practices/conventions/

**Comment by zubcevic** [comment]
I think I made all the requested changes. Except for an additional readme since the deployment documentation links to the Readme and each readme refers to the other

**Comment by fniessink** [comment]
Thank you!

**Pull request merged**